rustls/src/server/server_conn.rs

```diff
@@ -540,6 +545,58 @@ impl ServerConfig {
     }
 }
 
+/// A policy describing how invalid an invalid Server Name Indication (SNI) value from a client is handled by the server.
+///
+/// The only valid form of SNI according to relevant RFCs ([RFC6066], [RFC1035]) is
+/// non-IP-address host name, however some misconfigured clients may send a bare IP address, or
+/// another invalid value. Some servers may wish to ignore these invalid values instead of producing
+/// an error.
+///
+/// By default, Rustls will ignore invalid values that are an IP address (the most common misconfiguration)
+/// and error for all other invalid values.
+///
+/// When an SNI value is ignored, Rustls treats the client as if it sent no SNI at all.
+///
+/// [RFC1035]: https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1
+/// [RFC6066]: https://datatracker.ietf.org/doc/html/rfc6066#section-3
+#[derive(Default, Clone, Copy, PartialEq, Eq, Debug)]
+#[non_exhaustive]
+pub enum InvalidSniPolicy {
+    /// Reject all ClientHello messages that contain an invalid SNI value.
+    RejectAll,
+    /// Ignore an invalid SNI value in ClientHello messages if the value is an IP address.
+    ///
+    /// "Ignoring SNI" means accepting the ClientHello message, but acting as if the client sent no SNI.
+    #[default]
+    IgnoreIpAddresses,
+    /// Ignore all invalid SNI in ClientHello messages.
+    ///
+    /// "Ignoring SNI" means accepting the ClientHello message, but acting as if the client sent no SNI.
+    IgnoreAll,
+}
+
+impl InvalidSniPolicy {
+    /// Check if any potential invalid SNI in ClientHello is acceptable (i.e., ignorable)
+    /// by this policy.
+    ///
+    /// A `false` return value means there is an invalid SNI in ClientHello that is
+    /// rejected by this policy.
+    pub(super) fn is_acceptable(&self, client_hello: &ClientHelloPayload) -> bool {
```

Comment:
Nit: suggest simplifying this name to just `accept()`.

How about defining it more like this:

```rust
fn accept(&self, server_name: Option<&ServerNamePayload<'_>>) -> Result<Option<DnsName<'static>>, Error> { .. }
```

and moving the entire existing `match` into this function?

Reply:
I can do that. Doing that would require changing the type of `ServerConnectionData::sni` from `DnsName` to `ServerNamePayload`. Would you like me to do that?

Reply:
That seems like it might cause a bunch of pain in APIs that rely on this field...

Reply:
I think the solution here might be postponing storing into `ServerConnectionData` until after we've been able to call `InvalidSniPolicy::accept()`.